jerry-core/vm/vm.c

```diff
@@ -2194,7 +2192,52 @@ vm_loop (vm_frame_ctx_t *frame_ctx_p) /**< frame context */
         }
         case VM_OC_AWAIT:
         {
-          continue;
+          ecma_value_t promise = ecma_make_object_value (ecma_builtin_get (ECMA_BUILTIN_ID_PROMISE));
+          ecma_value_t argument = *(--stack_top_p);
+
+          result = ecma_promise_reject_or_resolve (promise, argument, true);
+          ecma_free_value (argument);
+
+          if (ECMA_IS_VALUE_ERROR (result))
+          {
+            goto error;
+          }
+
+          frame_ctx_p->call_operation = VM_EXEC_RETURN;
+          frame_ctx_p->byte_code_p = byte_code_p;
+          frame_ctx_p->stack_top_p = stack_top_p;
+
+          if (frame_ctx_p->block_result == ECMA_VALUE_UNDEFINED)
+          {
+            vm_executable_object_t *executable_object_p;
+            executable_object_p = opfunc_create_executable_object (frame_ctx_p, VM_CREATE_EXECUTABLE_OBJECT_ASYNC);
+
+            ecma_promise_async_then (result, ecma_make_object_value ((ecma_object_t *) executable_object_p));
+            ecma_deref_object ((ecma_object_t *) executable_object_p);
+            ecma_free_value (result);
+
+            ecma_object_t *old_new_target_p = JERRY_CONTEXT (current_new_target);
+            JERRY_CONTEXT (current_new_target) = ecma_builtin_get (ECMA_BUILTIN_ID_PROMISE);
+
+            result = ecma_op_create_promise_object (ECMA_VALUE_EMPTY, ECMA_PROMISE_EXECUTOR_EMPTY);
+
+            JERRY_ASSERT (ecma_is_value_object (result));
+            executable_object_p->frame_ctx.block_result = result;
+
+            JERRY_CONTEXT (current_new_target) = old_new_target_p;
+          }
+          else
+          {
+            uintptr_t diff = (uintptr_t) offsetof (vm_executable_object_t, frame_ctx);
```

Comment:
Maybe a better name for the `diff` would be `object_offset` and it can be a const value.

Reply:
(note: I changed this to const although the constant folding in the compiler does this automatically)